process/sync/shardForkDetector.go

```diff
@@ -69,51 +78,148 @@ func (sfd *shardForkDetector) AddHeader(
 }
 
-func (sfd *shardForkDetector) addFinalHeaders(finalHeaders []data.HeaderHandler, finalHeadersHashes [][]byte) {
-	finalCheckpointWasSet := false
-	for i := 0; i < len(finalHeaders); i++ {
-		isFinalHeaderNonceNotLowerThanCurrent := finalHeaders[i].GetNonce() >= sfd.finalCheckpoint().nonce
-		if isFinalHeaderNonceNotLowerThanCurrent {
-			if !finalCheckpointWasSet {
-				sfd.setFinalCheckpoint(&checkpointInfo{nonce: finalHeaders[i].GetNonce(), round: finalHeaders[i].GetRound()})
-				finalCheckpointWasSet = true
-			}
-
-			sfd.append(&headerInfo{
-				nonce: finalHeaders[i].GetNonce(),
-				round: finalHeaders[i].GetRound(),
-				hash:  finalHeadersHashes[i],
-				state: process.BHNotarized,
-			})
+// ReceivedSelfNotarizedHeaders is a registered call handler through which fork detector is notified when metachain
+// notarized new headers from self shard
+func (sfd *shardForkDetector) ReceivedSelfNotarizedHeaders(
+	shardID uint32,
+	selfNotarizedHeaders []data.HeaderHandler,
+	selfNotarizedHeadersHashes [][]byte,
+) {
+	if shardID != sharding.MetachainShardId || len(selfNotarizedHeaders) == 0 {
+		return
+	}
+
+	appended := sfd.appendSelfNotarizedHeaders(selfNotarizedHeaders, selfNotarizedHeadersHashes, shardID)
+	if appended {
+		sfd.computeFinalCheckpoint()
+	}
+}
+
+func (sfd *shardForkDetector) appendSelfNotarizedHeaders(
+	selfNotarizedHeaders []data.HeaderHandler,
+	selfNotarizedHeadersHashes [][]byte,
+	shardID uint32,
+) bool {
+
+	selfNotarizedHeaderAdded := false
+	finalNonce := sfd.finalCheckpoint().nonce
+
+	for i := 0; i < len(selfNotarizedHeaders); i++ {
+		if selfNotarizedHeaders[i].GetNonce() <= finalNonce {
+			continue
+		}
+
+		appended := sfd.append(&headerInfo{
+			nonce: selfNotarizedHeaders[i].GetNonce(),
+			round: selfNotarizedHeaders[i].GetRound(),
+			hash:  selfNotarizedHeadersHashes[i],
+			state: process.BHNotarized,
+		})
+		if appended {
+			log.Debug("added self notarized header in fork detector",
+				"shard", shardID,
+				"round", selfNotarizedHeaders[i].GetRound(),
+				"nonce", selfNotarizedHeaders[i].GetNonce(),
+				"hash", selfNotarizedHeadersHashes[i])
+
+			selfNotarizedHeaderAdded = true
 		}
 	}
+
+	return selfNotarizedHeaderAdded
+}
+
+func (sfd *shardForkDetector) computeFinalCheckpoint() {
+	sfd.mutHeaders.RLock()
+	for nonce, headersInfo := range sfd.headers {
+		if sfd.finalCheckpoint().nonce >= nonce {
+			continue
+		}
+
+		indexBHProcessed, indexBHNotarized := sfd.getProcessedAndNotarizedIndexes(headersInfo)
+		isProcessedBlockAlreadyNotarized := indexBHProcessed != -1 && indexBHNotarized != -1
+		if !isProcessedBlockAlreadyNotarized {
+			continue
+		}
+
+		sameHash := bytes.Equal(headersInfo[indexBHNotarized].hash, headersInfo[indexBHProcessed].hash)
+		if !sameHash {
+			continue
+		}
+
+		checkPointInfo := checkpointInfo{
+			nonce: nonce,
+			round: headersInfo[indexBHNotarized].round,
+			hash:  headersInfo[indexBHNotarized].hash,
+		}
+		sfd.setFinalCheckpoint(&checkPointInfo)
```

Comment:
shouldn't we set it only once - for the highest round header ?

Reply:
Yes, refactored